sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -199,10 +211,76 @@ case class InSubquery(values: Seq[Expression], query: ListQuery)
   override def children: Seq[Expression] = values :+ query
   override def nullable: Boolean = children.exists(_.nullable)
   override def foldable: Boolean = children.forall(_.foldable)
+}
+
+object SubqueryPredicate {
+
+  /**
+   * Only be used in SubqueryPredicate, and will be rewrite as Not(EqualTo(left, right)).
+   */
+  case class NotEqualTo(left: Expression, right: Expression) extends BinaryComparison {
+
+    override def symbol: String = "!="
+  }
+
+  /**
+   * Rewrite `NotEqualTo` as `Not(EqualTo)` and reserve others.
+   */
+  def getComparisonExpression[T <: BinaryComparison : ClassTag](
```

Comment:
To avoid this weird logic, IMO we'd better to define an actual binary comparison expr in `AstBuilder` as I left the comment in https://github.com/apache/spark/pull/23783#discussion_r256804080